### ISSUE TITLE
Bee eventually exceeds open files limit - 1.1.0-rc

### ISSUE THREAD

**Issue by ldeffenb** [ethersphere/bee#2399]
#### Summary
I have had several mainnet bee nodes terminate abruptly or stop working properly recently with "too many open files" errors in their log files.  I believe that the new Pinger which verifies peer connectivity via libp2p may be leaking sockets.  I am in the process of running a test with the current master with the hive pinger enabled and will do the same test with the pinger commented out. 

#### Steps to reproduce
Run a new node for a while and you'll likely run out of open files.  This seems to happen more with brand new nodes than with pre-existing nodes as the former has a lot more peers to learn.

#### Expected behavior
The number of open files should track the number of peer connections and number of levelDB open files, but the number of sockets seems to grow outside of the shown peer connections.

#### Actual behavior
Open files grows beyond expectations until it finally exceeds the system limit.

In an overnight test, my bee node went from 1794 file handles to 3379 file handles, while the connected peers only changed from 468 to 816.  And my db-open-files-limit is 200, although 1043 of the handles are .ldb files, 2370 are sockets which doesn't seem to track the 816 connected peers.

**Comment by ldeffenb** [comment]
Further, I just thought to check the metrics, and I find it hard to believe that this node has found ZERO unreachable peers out of 811,713 received.  I suspect that the Ping isn't even detecting what the Ping is designed to detect:

# HELP bee_hive_peers_handler_count Number of peer messages received.
# TYPE bee_hive_peers_handler_count counter
bee_hive_peers_handler_count 614312
# HELP bee_hive_peers_handler_peers_count Number of peers received in peer messages.
# TYPE bee_hive_peers_handler_peers_count counter
bee_hive_peers_handler_peers_count 811713
# HELP bee_hive_unreachable_peers_count Number of peers that are unreachable.
# TYPE bee_hive_unreachable_peers_count counter
bee_hive_unreachable_peers_count 0

**Comment by acud** [comment]
@aloknerurkar could you have a look?

**Comment by aloknerurkar** [comment]
@ldeffenb Went through the ping code and there does seem to be some issue. However `bee_hive_unreachable_peers_count ` at 0 is surprising. The issue I am looking into is the stream not closing cleanly.

We tried this change on our testnet for a few days and did not see this issue. I was hoping if you would be open to trying the fix on your nodes to verify? I'll share my branch.

**Comment by ldeffenb** [comment]
> We tried this change on our testnet for a few days and did not see this issue. I was hoping if you would be open to trying the fix on your nodes to verify? I'll share my branch.

Yes, I am 100% open to building and testing a different branch for this issue!

**Comment by vojtechsimetka** [comment]
I've encountered similar issue but with the blockchain connection on xDai. @aloknerurkar does your branch also tackle that? 

It takes ~30 mins to get this
```
ERRO[2021-08-09T10:20:56+02:00] failed syncing event listener, shutting down node err: Post "https://stake.getblock.io/mainnet/?api_key=CENSORED": dial tcp: lookup stake.getblock.io on 10.0.1.1:53: dial udp 10.0.1.1:53: socket: too many open files 
INFO[2021-08-09T10:20:56+02:00] api shutting down  
INFO[2021-08-09T10:20:56+02:00] greeting "SG pool" from peer: 5b3e1f25f7db475d4b9f6e4f389759451d6aa2b5a04213728bfa92f171f7782a 
INFO[2021-08-09T10:20:56+02:00] successfully connected to peer 55b23983ae05d989ae1fe336f2dc7251f932304f6bae5e3aa765c41e47efb9d0 (outbound) 
INFO[2021-08-09T10:20:56+02:00] pusher shutting down                         
INFO[2021-08-09T10:20:56+02:00] pull syncer shutting down                    
INFO[2021-08-09T10:20:56+02:00] kademlia shutting down                       
ERRO[2021-08-09T10:20:56+02:00] kademlia: could not broadcast to peer 55b23983ae05d989ae1fe336f2dc7251f932304f6bae5e3aa765c41e47efb9d0 
WARN[2021-08-09T10:20:56+02:00] could not send payment threshold announcement to peer 5b3e1f25f7db475d4b9f6e4f389759451d6aa2b5a04213728bfa92f171f7782a 
WARN[2021-08-09T10:20:56+02:00] peer not reachable when attempting to connect 
WARN[2021-08-09T10:20:56+02:00] peer not reachable when attempting to connect 
INFO[2021-08-09T10:20:56+02:00] kademlia persisting peer metrics             
zsh: killed     bee start --config ./bee.yaml
```

and a different version of the probably same error
```
ERRO[2021-08-09T08:50:04+02:00] failed syncing event listener, shutting down node err: put chain state: open bee-data-dir/statestore/000021.ldb: too many open files 
INFO[2021-08-09T08:50:04+02:00] api shutting down                            
INFO[2021-08-09T08:50:04+02:00] pusher shutting down                         
INFO[2021-08-09T08:50:04+02:00] pull syncer shutting down                    
INFO[2021-08-09T08:50:04+02:00] kademlia shutting down                       
INFO[2021-08-09T08:50:04+02:00] kademlia persisting peer metrics             
ERRO[2021-08-09T08:50:05+02:00] failed shutting down node: 1 error occurred:
	* statestore: open bee-data-dir/statestore/000021.ldb: too many open files
```

**Comment by aloknerurkar** [comment]
@ldeffenb Thanks. The branch is [pingfix.0](https://github.com/ethersphere/bee/tree/pingfix.0)

@vojtechsimetka This could manifest in different ways. Due to this bug using up the file descriptors, other operations can fail. Can you also try if the above branch works?

**Comment by ldeffenb** [comment]
Well, I can say that bee_hive_unreachable_peers_count is no longer zero!  Two nodes shortly after starting them with this branch:

bee_hive_peers_handler_peers_count 8602
bee_hive_unreachable_peers_count 1071

bee_hive_peers_handler_peers_count 6139
bee_hive_unreachable_peers_count 1135

I'll keep them running overnight and see what they look like in the AM.

**Comment by sonygod** [comment]
I did try to build and run this version,still not work! 

```haxe
time="2021-08-10T07:53:31Z" level=info msg="kademlia shutting down"
time="2021-08-10T07:53:36Z" level=warning msg="kademlia shutting down with announce goroutines"
time="2021-08-10T07:53:41Z" level=warning msg="kademlia manage loop did not shut down properly"
time="2021-08-10T07:53:41Z" level=info msg="kademlia persisting peer metrics"
```

**Comment by aloknerurkar** [comment]
@sonygod Can you update the branch and try once more?

**Comment by sonygod** [comment]
thank you @aloknerurkar  

**1.1.1-11c127d9**

,I just build and run that ,I'll report here again .

**Comment by ldeffenb** [comment]
> @sonygod Can you update the branch and try once more?

My nodes running this also died overnight.  Updated the branch, building, and will restart them shortly.  This time I'll try to watch the fd count via

```sudo ls -l /proc/54201/fd | grep socket | wc```

**Comment by sonygod** [comment]
new error here when update last branch.
@aloknerurkar  @ldeffenb 

```haxe 

time="2021-08-10T10:27:31Z" level=debug msg="kademlia: disconnected peer 142c40c26d26c67eb158201600cd035eb92603cefd8bbbd525e35b4e9181650a"
time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 142c40c26d26c67eb158201600cd035eb92603cefd8bbbd525e35b4e9181650a not found"
time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 14b85f988503cbcc6023ee4a941890cd39fb3b1135b855e7df5ef991c7d8ca61 not found"
time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 23ab7b674c619ec5e924c0d0b56296894be4e87df4a35459c53a07b12a119b82 not found"
time="2021-08-10T10:27:31Z" level=debug msg="kademlia: disconnected peer 15394b3af41eabe9eaf39fa15df8c948ed6e38cf2a444f09aa1b923543795bf5"
time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 15394b3af41eabe9eaf39fa15df8c948ed6e38cf2a444f09aa1b923543795bf5 not found"
time="2021-08-10T10:27:31Z" level=debug msg="kademlia: disconnected peer 14befc9168fa47e488a2da50076d5b560a60390d91a539d2eb4335dab1d88af4"
time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 14befc9168fa47e488a2da50076d5b560a60390d91a539d2eb4335dab1d88af4 not found"
time="2021-08-10T10:27:31Z" level=debug msg="kademlia: disconnected peer 14b978e3f912c0468323d249bae5e8788bba1f46f0a5aea7f58dd45ac618e2fa"
time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 14b978e3f912c0468323d249bae5e8788bba1f46f0a5aea7f58dd45ac618e2fa not found
```

**Comment by aloknerurkar** [comment]
> new error here when update last branch.
> @aloknerurkar @ldeffenb
> 
> ```hx
> time="2021-08-10T10:27:31Z" level=debug msg="kademlia: disconnected peer 142c40c26d26c67eb158201600cd035eb92603cefd8bbbd525e35b4e9181650a"
> time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 142c40c26d26c67eb158201600cd035eb92603cefd8bbbd525e35b4e9181650a not found"
> time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 14b85f988503cbcc6023ee4a941890cd39fb3b1135b855e7df5ef991c7d8ca61 not found"
> time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 23ab7b674c619ec5e924c0d0b56296894be4e87df4a35459c53a07b12a119b82 not found"
> time="2021-08-10T10:27:31Z" level=debug msg="kademlia: disconnected peer 15394b3af41eabe9eaf39fa15df8c948ed6e38cf2a444f09aa1b923543795bf5"
> time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 15394b3af41eabe9eaf39fa15df8c948ed6e38cf2a444f09aa1b923543795bf5 not found"
> time="2021-08-10T10:27:31Z" level=debug msg="kademlia: disconnected peer 14befc9168fa47e488a2da50076d5b560a60390d91a539d2eb4335dab1d88af4"
> time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 14befc9168fa47e488a2da50076d5b560a60390d91a539d2eb4335dab1d88af4 not found"
> time="2021-08-10T10:27:31Z" level=debug msg="kademlia: disconnected peer 14b978e3f912c0468323d249bae5e8788bba1f46f0a5aea7f58dd45ac618e2fa"
> time="2021-08-10T10:27:31Z" level=debug msg="libp2p disconnect: peer 14b978e3f912c0468323d249bae5e8788bba1f46f0a5aea7f58dd45ac618e2fa not found
> ```

This could be okay. The new changes actually do the disconnect now to free up the fd. Although these events should not be fired ideally as ping doesn't really use our main libp2p instance.

Ignoring this, lets see if the issue is fixed.

**Comment by ldeffenb** [comment]
@aloknerurkar Something is still obviously not working.  Did some probing with the output of sockstat and found the following in my logs as well as residual open sockets:

```bee@ubosm:/mnt/SSD/bee$ cat xdai6-Q.out | grep "112.18.156.63"
time="2021-08-10T07:42:00-04:00" level=debug msg="hive: peer fade1fc78a22cf97a9c14ff252bc067ef11091aaf6d30f977fd8fa2033433002: underlay /ip4/112.18.156.63/tcp/44180/p2p/16Uiu2HAmJvaxX4mn9EFuoHf5pfWT3MqBycECC4Q2uo5Az3CuR2Gb not reachable"
time="2021-08-10T08:00:17-04:00" level=debug msg="hive: peer c74c4ddc686a2511fbc9c6b642645ee97f5940fbc5829cac694fb1ef8bfff376: underlay /ip4/112.18.156.63/tcp/53894/p2p/16Uiu2HAmDfdKNcheL4RTsovKyF84J4PQSZGbpjxuAVKszVNZ7SfG not reachable"
time="2021-08-10T09:13:51-04:00" level=debug msg="hive: peer c74c4ddc686a2511fbc9c6b642645ee97f5940fbc5829cac694fb1ef8bfff376: underlay /ip4/112.18.156.63/tcp/43820/p2p/16Uiu2HAmDfdKNcheL4RTsovKyF84J4PQSZGbpjxuAVKszVNZ7SfG not reachable"
time="2021-08-10T10:44:31-04:00" level=debug msg="hive: peer 5404c3535ff2389ffc7fe0887c2d473ff41e87d105ec82c53a9b0c995452d3bf: underlay /ip4/112.18.156.63/tcp/33540/p2p/16Uiu2HAmP1P4r8WdhapA9KuFe8VUHh2Hysid5r6PuiCZJJ2KW5kA not reachable"
bee@ubosm:/mnt/SSD/bee$ sockstat | grep "112.18.156.63"
bee      bee-xdai-pingfi      120582   tcp4   192.168.10.20:11649       112.18.156.63:56420       ESTABLISHED
bee      bee-xdai-pingfi      120582   tcp4   192.168.10.20:11649       112.18.156.63:40482       ESTABLISHED
bee      bee-xdai-pingfi      120582   tcp4   192.168.10.20:11649       112.18.156.63:39530       ESTABLISHED
bee      bee-xdai-pingfi      120582   tcp4   192.168.10.20:11649       112.18.156.63:33736       ESTABLISHED
bee      bee-xdai-pingfi      120582   tcp4   192.168.10.20:11649       112.18.156.63:51522       ESTABLISHED
```
This was only one of the IP addresses I checked.  I might just write a tool that takes all current sockets' IP addresses and bounces them off of the current verbosity 5 log to see how many / which are actually the hive not reachable messages vs active p2p connections.

**Comment by ldeffenb** [comment]
> @ldeffenb the port numbers in the peer multiaddress are different... So the aforementioned lines might be a non-issue

I just searched the log file for each of those three peer IDs, and they only appear in the messages about those peers being non-reachable.  If they had connected in to my node, I should have seen them, or their IP addresses, in a successfully connected inbound message.  I do see these types of logs for other IP addresses that I've searched for from the sockstat listing.

I agree that the port number difference is a bit confusing being different than the underlay address port, but these sockets definitely seem to be related to the hive ping activity, IMHO.

**Comment by acud** [comment]
They might/probably be running the version which does not include the fix we are testing here. So they might have checked reachability of your node and never disposed of the socket (and connected to your node thereafter). Still makes sense and I'd file this under the non-issue drawer. The important bit would be to see that the _outgoing_ socket leak is gone.

**Comment by aloknerurkar** [comment]
@sonygod @ldeffenb I have created a PR for the current changes. Can you confirm this fixes the fd leak?

**Comment by sonygod** [comment]
No,you last version will crash my node .all my node was stop .

**Comment by acud** [comment]
@sonygod can you post the output of `ulimit -Sn`? `ulimit -Hn`?

**Comment by sonygod** [comment]
ulimit -n 60000

98 cores ,196 threads

**Comment by ldeffenb** [comment]
> @sonygod @ldeffenb I have created a PR for the current changes. Can you confirm this fixes the fd leak?

The 4 nodes that I have running this branch have survived for 24 hours now.  Better than they were before and the socket count has not increased nearly as dramatically as before.  The increase is likely as described above with non-fixed nodes holding their ping connection open to my node.

**Comment by sonygod** [comment]
I will check  and wait  next 12 hours. And report again

**Comment by ldeffenb** [comment]
> I set 5000 of open-file

You may improve your node uptime if you reduce that to 2000 or so.  At least for this test where you need the node still running to determine how many file descriptors are still in use.

**Comment by sonygod** [comment]
The 50 nodes that I have running this branch have survived for 12hours now ！ everthing seem ok and run well than before.

**Comment by acud** [comment]
@sonygod posting the last few log lines does not help - we need more information. usually posting the log lines that happened prior to the nodes existing would be helpful for us to debug the problem. also, if the node is shutting down it usually means that your blockchain backend is out of order, usually because the API limits (infura etc) are exhausted. If the node panics that's a different story.

I'm closing this issue as it should be fixed in #2403
